client/src/main/java/io/pravega/client/stream/impl/PendingEvent.java

```diff
@@ -38,26 +43,48 @@
      * The data to be written. Note this is limited to {@value #MAX_WRITE_SIZE} bytes.
      */
     private final ByteBuf data;
+    
+    private final int eventCount;
     /**
      * Callback to be invoked when the data is written.
      */
     private final CompletableFuture<Void> ackFuture;
        
-    private PendingEvent(String routingKey, ByteBuf data, CompletableFuture<Void> ackFuture) {
+    private PendingEvent(String routingKey, ByteBuf data, int eventCount, CompletableFuture<Void> ackFuture) {
         Preconditions.checkNotNull(data);
-        Preconditions.checkArgument(data.readableBytes() <= MAX_WRITE_SIZE, "Write size too large: %s", data.readableBytes());
         this.routingKey = routingKey;
         this.data = data;
         this.ackFuture = ackFuture;
+        this.eventCount = eventCount;
     }
     
     public static PendingEvent withHeader(String routingKey, ByteBuffer data, CompletableFuture<Void> ackFuture) {
+        ByteBuf eventBuf = getByteBuf(data);
+        return new PendingEvent(routingKey, eventBuf, 1, ackFuture);
+    }
+
+    private static ByteBuf getByteBuf(ByteBuffer data) {
         ByteBuf eventBuf = new Event(Unpooled.wrappedBuffer(data)).getAsByteBuf();
-        return new PendingEvent(routingKey, eventBuf, ackFuture);
-        
+        Preconditions.checkArgument(eventBuf.readableBytes() <= MAX_WRITE_SIZE, "Write size too large: %s", eventBuf.readableBytes());
+        return eventBuf;
     }
-    
+
+    public static PendingEvent withHeader(@NonNull String routingKey, @NonNull List<ByteBuffer> batch, @NonNull CompletableFuture<Void> ackFuture) {
+        Preconditions.checkArgument(!batch.isEmpty(), "Batch cannot be empty");
+        ByteBuf batchBuff = Unpooled.EMPTY_BUFFER;
+        for (int i = 0; i < batch.size(); i++) {
+            ByteBuf eventBuf = getByteBuf(batch.get(i));
+            batchBuff = Unpooled.wrappedUnmodifiableBuffer(batchBuff, eventBuf);
```

Comment:
Rather than making an N level deep structure, it would be better to add all the buffers into a list and then create a single wrapping buffer.